src/Arcus.WebApi.Unit/Security/Authentication/SharedAccessKeyAuthenticationAttributeTests.cs

```diff
@@ -127,14 +130,74 @@ public async Task AuthorizedRoute_WithSharedAccessKey_ShouldFailWithUnauthorized
+            {
+                // Assert
+                Assert.Equal(HttpStatusCode.Unauthorized, response.StatusCode);
+            }
+        }
+
+        [Fact]
+        public async Task AuthorizedRoute_WithSharedAccessKey__RegisteredWithSecretProvider_ShouldNotFailWithUnauthorized_WhenQuerystringValueMatchesSecretValue()
+        {
+            // Arrange
+            string secretValue = $"secret-{Guid.NewGuid()}";
+            _testServer.AddService<ISecretProvider>(new InMemorySecretProvider((SecretName, secretValue)));
+
+            // Act
+            using (HttpResponseMessage response = await SendAuthorizedHttpRequestWithQuerystring(ParameterName, secretValue))
+            {
+                // Assert
+                Assert.NotEqual(HttpStatusCode.Unauthorized, response.StatusCode);
+            }
+        }
+
+        [Fact]
+        public async Task AuthorizedRoute_WithSharedAccessKey_RegisteredWithCachedSecretProvider_ShouldNotFailWithUnauthorized_WhenQuerystringValueMatchesSecretValue()
+        {
+            // Arrange
+            string secretValue = $"secret-{Guid.NewGuid()}";
+            _testServer.AddService<ICachedSecretProvider>(new InMemorySecretProvider((SecretName, secretValue)));
+
+            // Act
+            using (HttpResponseMessage response = await SendAuthorizedHttpRequestWithQuerystring(ParameterName, secretValue))
+            {
+                // Assert
+                Assert.NotEqual(HttpStatusCode.Unauthorized, response.StatusCode);
+            }
+        }
+
+        [Fact]
+        public async Task AuthorizedRoute_WithSharedAccessKey_ShouldFailWithUnauthorized_WhenQuerystringValueNotMatchesSecretValue()
+        {
+            // Arrange
+            string secretValue = $"secret-{Guid.NewGuid()}";
+            _testServer.AddService<ISecretProvider>(new InMemorySecretProvider((SecretName, secretValue)));
+
+            // Act
+            using (HttpResponseMessage response = await SendAuthorizedHttpRequestWithQuerystring(ParameterName, $"something else then {nameof(secretValue)}"))
             {
                 // Assert
                 Assert.Equal(HttpStatusCode.Unauthorized, response.StatusCode);
             }
         }
 
+        [Fact]
+        public async Task AuthorizedRoute_WithSharedAccessKey_ShouldFailWithUnauthorized_WhenQuerystringParameterIsNotPresent()
+        {
+            // Arrange
+            string secretValue = $"secret-{Guid.NewGuid()}";
+            _testServer.AddService<ISecretProvider>(new InMemorySecretProvider((SecretName, secretValue)));
+
+            // Act
+            using (HttpResponseMessage response = await SendAuthorizedHttpRequestWithQuerystring($"something-else-then-{nameof(ParameterName)}", "some parameter value"))
+            {
+                // Assert
+                Assert.Equal(HttpStatusCode.Unauthorized, response.StatusCode);
+            }
+        }
+        
```

Comment:
I would like to  suggest the following tests:
1. Request should be unauthorized if I send both the header and query parameter, but header has wrong value
2. Request should be unauthorized if I send both the header and query parameter, but query parameter has wrong value
3. Request should be authorized if I send both the header and query parameter with a valid value

Reply:
Can you come up with more of them @stijnmoreels?

Reply:
| Request header | Request query | Authentication | Response |
| ----------------- | ---------------- | ---------------- | ----------- |
| invalid               | invalid              | header & query| Unauthorized |
| invalid               | valid                 | header & query| Unauthorized |   
| valid                 | invalid               | header & query| Unauthorized |
| valid                 | valid                  | header & query| Authorized |
| invalid               | invalid              | header | Unauthorized |
| invalid               | valid                 | header | Unauthorized |
| valid                 | invalid               | header | Authorized |
| valid                 | valid                  | header | Authorized|
| invalid               | invalid              | query   | Unauthorized |
| invalid               | valid                 | query   | Authorized |
| valid                  | invalid              | query   | Unauthorzed |
| valid                  | valid                 | query   | Authorized |

Reply:
Will add test for these cases

Reply:
> Request header	Request query	Authentication	Response
> invalid	invalid	header & query	Unauthorized
> invalid	valid	header & query	Unauthorized
> valid	invalid	header & query	Unauthorized
> valid	valid	header & query	Authorized
> invalid	invalid	header	Unauthorized
> invalid	valid	header	Unauthorized
> valid	invalid	header	Authorized
> valid	valid	header	Authorized
> invalid	invalid	query	Unauthorized
> invalid	valid	query	Authorized
> valid	invalid	query	Unauthorzed
> valid	valid	query	Authorized

@stijnmoreels these tests imply user has influence on kind of authorization performed, but this is imo not foreseen yet. Should this be added?

Reply:
@stijnmoreels @tomkerkhove Ignore the previous comment. Test cases have been updated

Reply:
Thanks!